File: benchmarks/cross-language/go/counting.go
// Go Counting Actor Benchmark (Savina-style)
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func getMessages() int {
	if env := os.Getenv("BENCHMARK_MESSAGES"); env != "" {
		if n, err := strconv.Atoi(env); err == nil {
			return n
		}
	}
	return 100000
}

var messages = getMessages()

func counter(inbox chan struct{}, done chan int) {
	count := 0
	for range inbox {
		count++
	}
	done <- count
}

func main() {
	fmt.Println("=== Go Counting Actor Benchmark ===")
	fmt.Printf("Messages: %d\n\n", messages)

	inbox := make(chan struct{}, 1024)
	done := make(chan int)

	go counter(inbox, done)

	start := time.Now()

	// Send all messages
	for i := 0; i < messages; i++ {
		inbox <- struct{}{}
	}
	close(inbox)

	// Wait for count
	count := <-done

	elapsed := time.Since(start)

	if count != messages {
		fmt.Printf("VALIDATION FAILED: expected %d, got %d\n", messages, count)
	}

	elapsedSec := elapsed.Seconds()
	throughput := float64(messages) / elapsedSec / 1e6
	nsPerMsg := elapsedSec * 1e9 / float64(messages)

	fmt.Printf("ns/msg:         %.2f\n", nsPerMsg)
	fmt.Printf("Throughput:     %.2f M msg/sec\n", throughput)
}

File: benchmarks/cross-language/rust/src/counting.rs
// Rust Counting Actor Benchmark (Savina-style)
use std::sync::mpsc::{channel, Sender, Receiver};
use std::thread;
use std::time::Instant;

fn get_messages() -> usize {
    std::env::var("BENCHMARK_MESSAGES")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(100_000)
}

fn counter(rx: Receiver<()>) -> usize {
    let mut count = 0;
    while rx.recv().is_ok() {
        count += 1;
    }
    count
}

fn main() {
    let messages = get_messages();
    println!("=== Rust Counting Actor Benchmark ===");
    println!("Messages: {}\n", messages);

    let (tx, rx) = channel();

    let handle = thread::spawn(move || counter(rx));

    let start = Instant::now();

    // Send all messages
    for _ in 0..messages {
        tx.send(()).unwrap();
    }
    drop(tx); // Signal done

    let count = handle.join().unwrap();

    let elapsed = start.elapsed().as_secs_f64();

    if count != messages {
        println!("VALIDATION FAILED: expected {}, got {}", messages, count);
    }

    let throughput = messages as f64 / elapsed / 1e6;
    let ns_per_msg = elapsed * 1e9 / messages as f64;

    println!("ns/msg:         {:.2}", ns_per_msg);
    println!("Throughput:     {:.2} M msg/sec", throughput);
}
